apps/tests/benchmarks/confidential.go

```diff
@@ -0,0 +1,205 @@
+	start := time.Now()
+	for n := 0; n < params.Concurrency; n++ {
+		wg.Add(1)
+		go func(chunk int) {
+			for i := parts * chunk; i < parts*(chunk+1); i++ {
+				// we use this to add a fake token to the cache.
+				// each token has a different scope which is what makes them unique
+				_, err := client.AuthResultFromToken(context.Background(), authParams, accesstokens.TokenResponse{
+					AccessToken:   accessToken,
+					ExpiresOn:     internalTime.DurationTime{T: time.Now().Add(1 * time.Hour)},
+					GrantedScopes: accesstokens.Scopes{Slice: []string{strconv.FormatInt(int64(i), 10)}},
+				}, true)
+				if err != nil {
+					panic(err)
+				}
+			}
+			wg.Done()
+		}(n)
+	}
+	wg.Wait()
+	return execTime{start: start, end: time.Now()}
+}
+
+func executeTest(client base.Client, params testParams) execTime {
+	wg := &sync.WaitGroup{}
+	fmt.Printf("Begin token retrieval.....")
+	start := time.Now()
+	for n := 0; n < params.Concurrency; n++ {
+		wg.Add(1)
+		go func() {
+			// retrieve each token once per goroutine
+			for tk := 0; tk < params.TokenCount; tk++ {
+				_, err := client.AcquireTokenSilent(context.Background(), base.AcquireTokenSilentParameters{
+					Scopes:      []string{strconv.FormatInt(int64(tk), 10)},
+					RequestType: accesstokens.ATConfidential,
+					Credential: &accesstokens.Credential{
+						Secret: "fake_secret",
+					},
+				})
+				if err != nil {
+					panic(err)
+				}
+			}
+			wg.Done()
+		}()
+	}
+	wg.Wait()
+	return execTime{start: start, end: time.Now()}
+}
+
+// Stats is used with statsTemplText for reporting purposes
+type Stats struct {
+	popExec     execTime
+	retExec     execTime
+	Concurrency int
+	Count       int64
+}
+
+// PopDur returns the total duration for populating the cache.
+func (s *Stats) PopDur() time.Duration {
+	return s.popExec.end.Sub(s.popExec.start)
+}
+
+// RetDur returns the total duration for retrieving tokens.
+func (s *Stats) RetDur() time.Duration {
+	return s.retExec.end.Sub(s.retExec.start)
+}
+
+// PopAvg returns the mean average of caching a token.
+func (s *Stats) PopAvg() time.Duration {
+	return s.PopDur() / time.Duration(s.Count)
+}
+
+// RetAvg returns the mean average of retrieving a token.
+func (s *Stats) RetAvg() time.Duration {
+	return s.RetDur() / time.Duration(s.Count)
+}
+
+var statsTemplText = `
+Test Results:
+[{{.Concurrency}} goroutines][{{.Count}} tokens] [population: total {{.PopDur}}, avg {{.PopAvg}}] [retrieval: total {{.RetDur}}, avg {{.RetAvg}}]
+==========================================================================
+`
+var statsTempl = template.Must(template.New("stats").Parse(statsTemplText))
+
+func main() {
+	tests := []testParams{
+		{
+			Concurrency: runtime.NumCPU(),
+			TokenCount:  100,
+		},
+		{
+			Concurrency: runtime.NumCPU(),
+			TokenCount:  1000,
+		},
+		{
+			Concurrency: runtime.NumCPU(),
+			TokenCount:  10000,
+		},
+		{
+			Concurrency: runtime.NumCPU(),
+			TokenCount:  20000,
```

Comment:
I know we wanted a larger upper bound but it's just too slow.  Inserting 50K tokens takes ~15 minutes on my machine.  Originally I had tried 100K but it took over an hour and upon token retrieval it failed because the dummy tokens had expired :(